internal/controller/rbac/definition/roles.go

```diff
@@ -99,6 +101,15 @@ func RenderClusterRoles(d *v1.CompositeResourceDefinition) []rbacv1.ClusterRole
 				Resources: []string{d.Spec.Names.Plural},
 				Verbs:     verbsEdit,
 			},
+			// The rule grants access to update the finalizers of resources that belong to the API group.
+			// If OwnerReferencesPermissionEnforcement is enabled, it allows after deleting the XRD to
+			// clean up resources that they no longer own, so that the system can be kept in a consistent state.
```

Comment:
It's probably not the XRD (the definition) that is being deleted, but rather the XR, right? I think the intention here is to allow the Crossplane pod to add a finalizer to XRs, so that it can create composed resources that block their deletion.

If that is the case, it's not clear to me why we need this rule at all. I see you observed issues with composition per https://github.com/crossplane/crossplane/issues/3443#issuecomment-1306828892, but I don't understand why that is happening. Crossplane already has `verbsEdit` (i.e. `*`) access to XRs, which I would assume included access to update XR finalizers. Perhaps `*` not apply to the `/finalizers` resource?

Crossplane definitely has RBAC access to add/remove XR finalizers today - see for example https://github.com/crossplane/crossplane/blob/0951db/internal/controller/apiextensions/composite/reconciler.go#L457

If we do need this rule, I'm pretty sure it should be on the `system` role rather than the `edit` role. The `system` role is what the Crossplane pod runs as - `edit` is for human users. I'd also prefer the comment to be a little more detailed, for example:

> Crossplane reconciles an XR by creating one or more composed resources. These composed resources are controlled (in the owner reference sense) by the XR. Crossplane needs permission to set finalizers on XRs in order to create resources that block their deletion when the OwnerReferencesPermissionEnforcement admission controller is enabled.

Reply:
> I don't understand why that is happening. Crossplane already has verbsEdit (i.e. *) access to XRs, which I would assume included access to update XR finalizers. Perhaps * not apply to the /finalizers resource?

I see you've moved this rule to `system` - thank you. It's still not clear to me why/whether it's needed though per my above comment. Can you clarify?

Reply:
@negz yes,

When XR tries to create a managed resource with ownerReference them, these rules give to allow update finalizers to MR

without this permission, crossplane cannot create resources. I've got issues when I try to use my own XRD:
`cannot compose resources: cannot apply composed resource: cannot create
        object: buckets.s3.aws.crossplane.io "default-bucket-test1" is
        forbidden: cannot set blockOwnerDeletion if an ownerReference refers to
        a resource you can't set finalizers on: , <nil>`
this exception happened is here:
https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/composite/composition_pt.go#L244

and this is when I try to use  
`kubectl apply -f https://raw.githubusercontent.com/crossplane/crossplane/release-1.10/docs/snippets/compose/claim-aws.yaml`
error from XR:
`cannot publish connection details: cannot apply connection secret:
        cannot create object: secrets "" is
        forbidden: cannot set blockOwnerDeletion if an ownerReference refers to
        a resource you can't set finalizers on: , <nil>`

this exception happened is here:
https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/composite/reconciler.go#L533

p.s. s, you're absolutely right about set Finalizers in any resource like this https://github.com/crossplane/crossplane/blob/0951db/internal/controller/apiextensions/composite/reconciler.go#L457, maybe I should research how crossplane works with RBAC and resource, but this is needed more time for me. I noticed between Crossplane and any Providers have a gap in permissions

Reply:
Interesting. I guess it needs update on `resource`/finalizers` specifically. I would have thought that was implied by having `*` on `resource`. Thanks for checking!